packages/catalog-server/src/lib/catalog.ts

```diff
@@ -31,8 +53,111 @@ export class Catalog {
     this.#repository = init.repository;
   }
 
+  /**
+   * Imports a package, which inserts or updates the package-wide metadata like
+   * dist-tags, versions, and times.
+   *
+   * If the dist-tags have changed since the last import, package versions will
+   * be updated with the updated dist-tags.
+   *
+   * If the 'latest' dist-tag needs importing, that package version will be
+   * imported too.
+   */
+  async importPackage(
+    packageName: string,
+    packageRefreshInterval = defaultPackageRefreshInterval
+  ) {
+    const currentPackageInfo = await this.#repository.getPackageInfo(
+      packageName
+    );
+
+    if (currentPackageInfo === undefined) {
+      // For an initial import, mark the package as initializing:
+      await this.#repository.startPackageImport(packageName);
+    } else {
+      // Check that we haven't imported recently, and abort if we have:
+      const updatedDate = currentPackageInfo.lastUpdate;
+      const updatedInstant = toTemporalInstant(updatedDate);
+      const now = Temporal.Now.instant();
+      const timeSinceUpdate = now.since(updatedInstant);
+      if (
+        Temporal.Duration.compare(timeSinceUpdate, packageRefreshInterval) ===
+        -1
+      ) {
+        return;
+      }
+      await this.#repository.startPackageUpdate(packageName);
+    }
+
+    // Fetch package metadata from npm:
+    let newPackage: Package | undefined;
+    try {
+      newPackage = await this.#files.getPackageMetadata(packageName);
+    } catch (e) {
+      await this.#repository.endPackageImportWithError(packageName);
+      return;
+    }
+
+    if (newPackage === undefined) {
+      await this.#repository.endPackageImportWithNotFound(packageName);
+      // TODO (justinfagnani): a crazy edge case would be a package that was
```

Comment:
This isn't always a crazy error, some systems handle some internal errors by giving a 404 or similar rather than a 500 error

Hopefully the npm registry isn't one, but something to be mindful of

Reply:
That is true. I'll look into it